File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/MassStorageSample/src/com/hp/workpath/sample/massstoragesample/task/ReadFileTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.massstoragesample.task;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.hp.workpath.api.massstorage.CustomerDataFile;
import com.hp.workpath.api.massstorage.CustomerDataFileUtils;
import com.hp.workpath.sample.massstoragesample.Logger;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ReadFileTask {
    private final WeakReference<Context> mContextRef;
    private CustomerDataFile mCustomerDataFile;
    private ReadFileTaskInterface readFileTaskInterface;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public ReadFileTask(Context context, CustomerDataFile customerDataFile, ReadFileTaskInterface readFileTaskInterface) {
        this.mContextRef = new WeakReference<>(context);
        this.mCustomerDataFile = customerDataFile;
        this.readFileTaskInterface = readFileTaskInterface;
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    String content = "";
                    InputStream fis = null;
                    BufferedInputStream bis = null;
                    long fileLength = mCustomerDataFile.length();

                    try {
                        int bufferSize = 64;
                        if (fileLength < bufferSize) {
                            bufferSize = (int) fileLength;
                        }

                        byte[] buffer = new byte[bufferSize];
                        fis = CustomerDataFileUtils.openInputStream(mContextRef.get(), mCustomerDataFile);
                        bis = new BufferedInputStream(fis);
                        bis.read(buffer);
                        content = new String(buffer);
                    } catch (Throwable t) {
                        Logger.showResult(null, "CustomerDataFileUtils.openInputStream " + t.getMessage());
                    } finally {
                        if (bis != null) {
                            try {
                                bis.close();
                            } catch (IOException e) {
                            }
                        }
                        if (fis != null) {
                            try {
                                fis.close();
                            } catch (IOException e) {
                            }
                        }
                    }
                    onPostExecute(content);
                }
            });
        } catch (Exception e) {
            Logger.showResult(null, "CustomerDataFileUtils.openInputStream " + e.getMessage());
            executor.shutdown();
        }
    }

    public interface ReadFileTaskInterface {
        void fileContent(String content);
    }

    private void onPostExecute(String content) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                readFileTaskInterface.fileContent(content);
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/MassStorageSample/src/com/hp/workpath/sample/massstoragesample/task/ReadFileTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.massstoragesample.task

import android.content.Context
import com.hp.workpath.api.massstorage.CustomerDataFile
import com.hp.workpath.api.massstorage.CustomerDataFileUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.lang.ref.WeakReference

class ReadFileTask(context: Context, customerDataFile: CustomerDataFile, private val readFileTaskInterface: ReadFileTaskInterface) {
    private val mContextRef: WeakReference<Context> = WeakReference(context)
    private val mCustomerDataFile: CustomerDataFile = customerDataFile
    private var mThrowable: Throwable? = null

    interface ReadFileTaskInterface {
        fun fileContent(content: String?)
    }

    suspend fun execute() {
        var content = ""
        var fis: InputStream? = null
        var bis: BufferedInputStream? = null
        val fileLength = mCustomerDataFile.length()
        try {
            var bufferSize = 64
            if (fileLength < bufferSize) {
                bufferSize = fileLength.toInt()
            }
            val buffer = ByteArray(bufferSize)
            fis = CustomerDataFileUtils.openInputStream(mContextRef.get(), mCustomerDataFile)
            bis = BufferedInputStream(fis)
            bis.read(buffer)
            content = String(buffer)
        } catch (t: Throwable) {
            mThrowable = t
        } finally {
            if (bis != null) {
                try {
                    bis.close()
                } catch (e: IOException) {
                }
            }
            if (fis != null) {
                try {
                    fis.close()
                } catch (e: IOException) {
                }
            }
        }
        onPostExecute(content)
    }

    private suspend fun onPostExecute(content: String) {
        withContext(Dispatchers.Main) {
            if (mThrowable != null) {
                readFileTaskInterface.fileContent("CustomerDataFileUtils.openInputStream ${mThrowable?.message}")
            } else {
                readFileTaskInterface.fileContent(content)
            }
        }
    }
}